# tensorflow-GPU 사용법

### 아래 사이트를 참고하였습니다.

https://www.pugetsystems.com/labs/hpc/The-Best-Way-to-Install-TensorFlow-with-GPU-Support-on-Windows-10-Without-Installing-CUDA-1187/

![image](https://user-images.githubusercontent.com/42320146/54493844-c2a24280-4917-11e9-9dbe-ae0a5c46d338.png)


In [1]:
import tensorflow as tf

In [2]:
hello = tf.constant('hello, Tensorflow!')
sess= tf.Session()
print(sess.run(hello))

b'hello, Tensorflow!'


# keras-GPU 사용법

![image](https://user-images.githubusercontent.com/42320146/54493954-b5d21e80-4918-11e9-82ef-f9c3d5f780f5.png)


![image](https://user-images.githubusercontent.com/42320146/54493971-d39f8380-4918-11e9-908a-239a824a09d1.png)


![image](https://user-images.githubusercontent.com/42320146/54493981-efa32500-4918-11e9-9de1-67cdea389272.png)


![image](https://user-images.githubusercontent.com/42320146/54493992-03e72200-4919-11e9-839d-34b0efa6b045.png)


![image](https://user-images.githubusercontent.com/42320146/54494003-195c4c00-4919-11e9-8360-48b570d6e79e.png)


![image](https://user-images.githubusercontent.com/42320146/54494010-355fed80-4919-11e9-806f-dc3c96678722.png)


In [4]:
import numpy as np
import time
def nowtime(past):
    now = time.time()
    print(now)
    print("period[second] : {}".format(now - past) )
    return now
from keras.datasets import mnist

((X_train, y_train), (X_test, y_test)) = mnist.load_data()

Using TensorFlow backend.


11493376/11490434 [==============================] - 9s 1us/step


In [10]:
def sigmoid(n):
    return 1 / (1 + np.exp(-n))

In [5]:
X_train = X_train.reshape(60000, 28 * 28)
X_test = X_test.reshape(10000, 28 * 28)

print(X_train.shape, X_test.shape)

(60000, 784) (10000, 784)


In [6]:
y_train_hot = np.eye(10)[y_train]

print(y_train[0:5])
y_train_hot[0:5]

[5 0 4 1 9]


array([[0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]])

In [7]:
y_test_hot = np.eye(10)[y_test]

print(y_test[0:5])
y_test_hot[0:5]

[7 2 1 0 4]


array([[0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.]])

In [11]:
pasttime=0

pasttime = nowtime(pasttime)

learning_rate = 0.00001

w1 = np.random.uniform(low=-0.058, high=+0.058, size=(784, 1000))
w2 = np.random.uniform(low=-0.077, high=+0.077, size=(1000, 10))

b1 = np.random.uniform(low=0, high=0, size=(1, 1000))
b2 = np.random.uniform(low=0, high=0, size=(1, 10))

num_epoch = 1000


for epoch in range(num_epoch):
  # Forward propagation
  l1 = X_train.dot(w1) + b1
  o1 = sigmoid(l1)
  l2 = o1.dot(w2) + b2
  o2 = sigmoid(l2)
  
  predict = np.argmax(o2, axis=1)
  error = (predict != y_train).mean()
  if error < 0.1:
    break
    
  if epoch % 10 == 0:
    print('{0:3} error = {1:.5f}'.format(epoch, error))
    print('{0:3} w1(mean) = {1:.5f}, w1(std) = {2:.5f}'.format(epoch, w1.mean(), w1.std()))
    print('{0:3} w2(mean) = {1:.5f}, w2(std) = {2:.5f}'.format(epoch, w2.mean(), w2.std()))
    print('{0:3} b1(mean) = {1:.5f}, b1(std) = {2:.5f}'.format(epoch, b1.mean(), b1.std()))
    print('{0:3} b2(mean) = {1:.5f}, b2(std) = {2:.5f}'.format(epoch, b2.mean(), b2.std()))
    
  # Backpropagation
  d2 = o2 - y_train_hot
  d1 = d2.dot(w2.T) * o1 * (1 - o1)
  
  w2 = w2 -learning_rate * o1.T.dot(d2)
  b2 = b2 -learning_rate * d2.mean(axis=0)
  w1 = w1 -learning_rate * X_train.T.dot(d1)
  b1 = b1 -learning_rate * d1.mean(axis=0)
 


print('{0:3} error = {1:.5f}'.format(epoch, error))
  
pasttime = nowtime(pasttime)

1552840096.1883516
period[second] : 1552840096.1883516
  0 error = 0.90862
  0 w1(mean) = -0.00000, w1(std) = 0.03350
  0 w2(mean) = 0.00018, w2(std) = 0.04453
  0 b1(mean) = 0.00000, b1(std) = 0.00000
  0 b2(mean) = 0.00000, b2(std) = 0.00000
 10 error = 0.85825
 10 w1(mean) = -0.00473, w1(std) = 0.03743
 10 w2(mean) = -0.11769, w2(std) = 0.12777
 10 b1(mean) = -0.00000, b1(std) = 0.00000
 10 b2(mean) = -0.00000, b2(std) = 0.00000
 20 error = 0.90128
 20 w1(mean) = -0.00523, w1(std) = 0.03783
 20 w2(mean) = -0.13622, w2(std) = 0.14184
 20 b1(mean) = -0.00000, b1(std) = 0.00000
 20 b2(mean) = -0.00001, b2(std) = 0.00000
 30 error = 0.90125
 30 w1(mean) = -0.00545, w1(std) = 0.03811
 30 w2(mean) = -0.13316, w2(std) = 0.14697
 30 b1(mean) = -0.00000, b1(std) = 0.00000
 30 b2(mean) = -0.00001, b2(std) = 0.00000
 40 error = 0.79102
 40 w1(mean) = -0.00544, w1(std) = 0.03834
 40 w2(mean) = -0.12797, w2(std) = 0.15056
 40 b1(mean) = -0.00000, b1(std) = 0.00000
 40 b2(mean) = -0.00000, b2(std

In [12]:
import pandas as pd

l1 = X_test.dot(w1)
o1 = sigmoid(l1)
l2 = o1.dot(w2)
o2 = sigmoid(l2)

In [14]:
predict = np.argmax(o2, axis=1)

accuracy = (y_test == predict).mean()
print('Accuracy = {0:.5f}'.format(accuracy))

pd.DataFrame({'y(actual)' : y_test, 'y(predict)' : predict}).head(15)

Accuracy = 0.90930


TypeError: __init__() got an unexpected keyword argument 'max_rows'

    y(actual)  y(predict)
0           7           7
1           2           2
2           1           1
3           0           0
4           4           4
5           1           1
6           4           4
7           9           9
8           5           6
9           9           9
10          0           0
11          6           6
12          9           9
13          0           0
14          1           1

learning_rate = 0.00001

w1 = np.random.uniform(low=-0.058, high=+0.058, size=(784, 1000)) w2 = np.random.uniform(low=-0.077, high=+0.077, size=(1000, 10))

b1 = np.random.uniform(low=0, high=0, size=(1, 1000)) b2 = np.random.uniform(low=0, high=0, size=(1, 10))

num_epoch = 1000

### train 데이터로 모델링시 -> epoch : 194 error = 0.09142
### 트레이닝 시간 :16.7m

*모델에 test 데이터 적용시 Accuracy = 0.90930

# => 애폭이 줄어 절대적 비교가 되진 않지만 시간 단축을 유추해 볼 수 있다.



# vs

## Colab TPU

learning_rate = 0.00001

w1 = np.random.uniform(low=-0.058, high=+0.058, size=(784, 1000)) w2 = np.random.uniform(low=-0.077, high=+0.077, size=(1000, 10))

b1 = np.random.uniform(low=0, high=0, size=(1, 1000)) b2 = np.random.uniform(low=0, high=0, size=(1, 10))

num_epoch = 1000

### train 데이터로 모델링시 -> epoch : 256 error = 0.09170
### 트레이닝 시간 :44m

*모델에 test 데이터 적용시 Accuracy = 0.90940